src/app/tests/TestConcreteAttributePath.cpp

```diff
@@ -0,0 +1,119 @@
+{
+    ConcreteAttributePath path_one(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3);
+    ConcreteAttributePath path_two(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3);
+    NL_TEST_ASSERT(aSuite, path_one == path_two);
+}
+
+void TestConcreteAttributePathInequalityDifferentAttributeId(nlTestSuite * aSuite, void * aContext)
+{
+    ConcreteAttributePath path_one(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3);
+    ConcreteAttributePath path_two(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/4);
+    NL_TEST_ASSERT(aSuite, path_one != path_two);
+}
+
+void TestConcreteReadAttributePathEqualityDefaultConstructor(nlTestSuite* aSuite, void * aContext) {
+    ConcreteReadAttributePath path_one;
+    ConcreteReadAttributePath path_two;
+    NL_TEST_ASSERT(aSuite, path_one == path_two);
+}
+
+void TestConcreteReadAttributePathInequalityDifferentListIndex(nlTestSuite* aSuite, void * aContext) {
+    ConcreteReadAttributePath path_one(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, /*aListIndex=*/4);
+    ConcreteReadAttributePath path_two(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, /*aListIndex=*/5);
+    NL_TEST_ASSERT(aSuite, path_one != path_two);
+}
+
+void TestConcreteDataAttributePathEqualityDefaultConstructor(nlTestSuite* aSuite, void * aContext) {
+    ConcreteDataAttributePath path_one;
+    ConcreteDataAttributePath path_two;
+    NL_TEST_ASSERT(aSuite, path_one == path_two);
+}
+
+void TestConcreteDataAttributePathInequalityDifferentDataVersion(nlTestSuite* aSuite, void * aContext) {
+    ConcreteDataAttributePath path_one(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, /*aDataVersion=*/MakeOptional(4));
+    ConcreteDataAttributePath path_two(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, /*aDataVersion=*/MakeOptional(5));
+    NL_TEST_ASSERT(aSuite, path_one != path_two);
+}
+
+void TestConcreteDataAttributePathInequalityDifferentListOp(nlTestSuite* aSuite, void * aContext) {
+    ConcreteDataAttributePath path_one(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, ConcreteDataAttributePath::ListOperation::ReplaceAll, /*aListIndex=*/4);
+    ConcreteDataAttributePath path_two(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, ConcreteDataAttributePath::ListOperation::ReplaceItem, /*aListIndex=*/4);
+    NL_TEST_ASSERT(aSuite, path_one != path_two);
+}
+
+void TestConcreteDataAttributePathInequalityDifferentListIndex(nlTestSuite* aSuite, void * aContext) {
+    ConcreteDataAttributePath path_one(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, ConcreteDataAttributePath::ListOperation::NotList, /*aListIndex=*/4);
+    ConcreteDataAttributePath path_two(/*aEndpointId=*/1, /*aClusterId=*/2, /*aAttributeId=*/3, ConcreteDataAttributePath::ListOperation::NotList, /*aListIndex=*/5);
+    NL_TEST_ASSERT(aSuite, path_one != path_two);
+}
+
+} // anonymous namespace
+
+namespace {
```

Comment:
you can just not close above - seems not need to close and re-open the anon namespace.

Reply:
Done